File: practice/convert/example2_book_library/book-library/src/main/java/dev/starryeye/book_library/domain/book/Book.java
package dev.starryeye.book_library.domain.book;

import dev.starryeye.book_library.domain.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@Entity
@Table(name = "books")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(value = AuditingEntityListener.class)
public class Book extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String bookname;

    private Long loanedUserId;

    @Version
    private Long version;

    @Builder
    private Book(Long id, String bookname, Long loanedUserId, Long version) {
        if (bookname == null || bookname.isBlank()) {
            throw new IllegalArgumentException("bookname must not be blank");
        }

        this.id = id;
        this.bookname = bookname;
        this.loanedUserId = loanedUserId;
        this.version = version;
    }

    public static Book createBook(String bookname) {
        return Book.builder()
                .id(null)
                .bookname(bookname)
                .loanedUserId(null)
                .version(0L)
                .build();
    }

    public void loanTo(Long userId) {
        if (userId == null) {
            throw new IllegalArgumentException("userId must not be null");
        }
        if (loanedUserId != null) {
            throw new IllegalStateException("book is already loaned, id = " + id);
        }

        this.loanedUserId = userId;
    }
}

File: practice/convert/example2_book_library/kotlin/book-library/src/main/kotlin/dev/starryeye/book_library/domain/book/Book.kt
package dev.starryeye.book_library.domain.book

import dev.starryeye.book_library.domain.BaseEntity
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.springframework.data.jpa.domain.support.AuditingEntityListener

@Entity
@Table(name = "books")
@EntityListeners(AuditingEntityListener::class)
class Book(
    val bookname: String,
    val category: String,
) : BaseEntity() {

    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    init {
        if (bookname.isBlank()) {
            throw IllegalArgumentException("Book name cannot be null or blank")
        }
    }

    // object mother 패턴..
    // 해당 객체(Book)의 프로퍼티가 늘어나더라도 Test code 에 전파되지 않는다.
    //      test code 에서는 Book 객체 생성시 Book.fixture() 만을 사용
    // 참고. DTO 는 Domain 객체보다 많이 사용되지 않는 편이라 fixture 를 안만드는 편
    companion object {
        fun fixture(
            bookname: String = "book name",
            category: String = "book category",
        ): Book {
            return Book(
                bookname = bookname,
                category = category,
            )
        }
    }
}